routers/repo/branch.go

```diff
@@ -30,3 +32,48 @@ func Branches(ctx *context.Context) {
 	ctx.Data["Branches"] = brs
 	ctx.HTML(200, tplBranch)
 }
+
+// CreateBranch creates new branch in repository
+func CreateBranch(ctx *context.Context, form auth.NewBranchForm) {
+	if !ctx.Repo.CanCreateBranch() {
+		ctx.Handle(404, "CreateBranch", nil)
+	}
+
+	if len(form.NewBranchName) == 0 || validation.GitRefNamePattern.MatchString(form.NewBranchName) {
+		ctx.Flash.Error(ctx.Tr("form.NewBranchName") + ctx.Tr("form.git_ref_name_error"))
+		ctx.Redirect(ctx.Repo.RepoLink + "/src/" + ctx.Repo.BranchName)
+		return
+	}
+
+	if len(form.NewBranchName) > 100 {
```

Comment:
It was @ethantkoenig suggestion. But I'll change it to 255 to be same as GitHub's limit